fix issue/5252 - CSVインポート時にデリミタがエスケープされていることがある

TPS-NghiaLe opened the pull request:
<!-- 以下を参考にコメントを作成してください。 -->

## 概要(Overview・Refs Issue)
<!-- PullRequestの目的、関連するIssue番号など -->
https://github.com/EC-CUBE/ec-cube/issues/5252 の修正

## 方針(Policy)
<!-- このPullRequestを作るにあたって考慮したものや除外した内容
  - 例）Symfony のXXにならって作成、3系で同様の仕様があったため など -->
Shift_JISの特殊文字により、ファイル内のデータが正しくない取得されて、エンコードが不正になります。

解決方法 : したがって、コントローラーで処理する前にエンコードデータをと取得することが必要です。

## 実装に関する補足(Appendix)
<!-- コードだけではわかりづらい点など、実装するにあたってレビューアに追加で伝えておきたいこと -->

## テスト（Test)
<!-- テストを行っている範囲など、レビューアが安心できるような情報 -->

## 相談（Discussion）
<!-- 相談したいことや意見を求めたいこと -->

## マイナーバージョン互換性保持のための制限事項チェックリスト
<!-- マイナーバージョンでは、機能・プラグイン・デザインテンプレート互換性を損なう変更は原則取り込みません。 -->

- [ ] 既存機能の仕様変更はありません
- [ ] フックポイントの呼び出しタイミングの変更はありません
- [ ] フックポイントのパラメータの削除・データ型の変更はありません
- [ ] twigファイルに渡しているパラメータの削除・データ型の変更はありません
- [ ] Serviceクラスの公開関数の、引数の削除・データ型の変更はありません
- [ ] 入出力ファイル(CSVなど)のフォーマット変更はありません

## レビュワー確認項目

- [ ] 動作確認
- [ ] コードレビュー
- [ ] E2E/Unit テスト確認(テストの追加・変更が必要かどうか)
- [ ] 互換性が保持されているか
- [ ] セキュリティ上の問題がないか
  - [ ] 権限を超えた操作が可能にならないか
  - [ ] 不要なファイルアップロードがないか
  - [ ] 外部へ公開されるファイルや機能の追加ではないか
  - [ ] テンプレートでのエスケープ漏れがないか

TPS-NghiaLe committed:
fix issue/5252 - CSVインポート時にデリミタがエスケープされていることがある

nanasess committed:
Windows に限定しないよう修正

nanasess committed:
ユニットテスト追加

see https://github.com/EC-CUBE/ec-cube/issues/5252

nanasess reviewed (changes requested):
ロケールの関係？で、正しく CSV ファイルをパースできなくなる模様。列が欠落したりするため要調査

nanasess commented:
> ロケールの関係？で、正しく CSV ファイルをパースできなくなる模様。列が欠落したりするため要調査

`SplFileObject::READ_CSV` が SJIS に対応していないため、正常にパースできない。
2系のように、 `stream_filter` を使用することで実現できるかもしれない
https://github.com/EC-CUBE/ec-cube2/blob/3129289ea4629337946609025cf3e3cb96436000/data/class/helper/SC_Helper_CSV.php#L380-L400

nanasess committed:
Merge remote-tracking branch 'upstream/4.2' into issues/5252

nanasess committed:
Stream filter で文字エンコーディングを変換する

- file_get_contents でファイルを読み込んで変換していたのを stream filter で変換するよう修正
- 改行コード変換も stream filter で変換するよう修正

nanasess commented:
stream filter を使用して変換することで対応しました。
また、 `AbstractCsvImportController` で処理していた改行コード変換も stream filter を使用するようにしました

chihiro-adachi commented:
@nanasess 
ありがとうございます。

こちらのCSV、商品CSV登録時にエラーとなりました。見ていただくことはできますでしょうか？
[product_20220715141042.csv](https://github.com/EC-CUBE/ec-cube/files/9118008/product_20220715141042.csv)

<img width="482" alt="image" src="https://user-images.githubusercontent.com/8196725/179155413-8d502ae8-7e0f-4bc0-b7d4-7e9c55293fe6.png">

nanasess commented:
@chihiro-adachi 

> こちらのCSV、商品CSV登録時にエラーとなりました。見ていただくことはできますでしょうか？
> [product_20220715141042.csv](https://github.com/EC-CUBE/ec-cube/files/9118008/product_20220715141042.csv)

このCSVファイルには商品IDの項目が無いためエラーになっていたようです。商品IDを追加したところ正常にアップロードできましたのでご確認ください

chihiro-adachi commented:
商品IDを追加したCSVでも同様のエラーとなりました。
ファイルを添付しておきます。

[product-sjis.csv](https://github.com/EC-CUBE/ec-cube/files/9118496/product-sjis.csv)
[product-utf-8.csv](https://github.com/EC-CUBE/ec-cube/files/9118495/product-utf-8.csv)

nanasess commented:
@chihiro-adachi stream filter に関係なく、 `$form['import_file']->getData()`  でデータを取り出した時点で以下のようにデータが壊れてしまう模様です。。。

```
商品ID,公開ステータス(ID),販売種別(ID),商品名,販売価格,在庫数,在庫数無制限フラグ
,1,1,"10テスト機構",2800,100,
,1,1,11テスト機構"",2800,100,
```

chihiro-adachi committed:
ファイル分割前にUTF-8変換をかける

chihiro-adachi commented:
@nanasess 

ファイルを分割するタイミングで壊れてしまうようなので、分割前にフィルタかける必要がありそうです。

一応以下でうまくいきました。
https://github.com/chihiro-adachi/ec-cube/commit/0bf4a28bf919a9a99409de7dc10601ccf79b1974

nanasess commented:
@chihiro-adachi なるほど、ありがとうございます！

nanasess committed:
Merge pull request #58 from chihiro-adachi/issues/5252

ファイル分割前にUTF-8変換をかける

nanasess committed:
リファクタリング

nanasess committed:
CsvImportService::convertEncodingRows() は非推奨

nanasess commented:
@chihiro-adachi いただいたコミットを取り込んでリファクタリングしました。これで大丈夫だと思います。

chihiro-adachi commented:
@nanasess 
ありがとうございます！
sjis.csv, utf-8.csvともに正常に登録できました。
